### PR TITLE
C++ compilers: Be sane with standard libraries

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This is long overdue, @LnL7 Time for you to say "I told you so!" :).

CC @matthewbauer @TravisWhitaker

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built stdenv on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Seeing that this is such a massive rebuild, on I am all ears on how we can do more nix and less bash to make the inevitable follow-up work easier.

Maybe we could purely concat lists for all those `*flags` file and write once? The building detection the wrapper script does currently I am not too pleased with.

File: pkgs/development/compilers/llvm/7/default.nix [hunk 1]
```diff
@@ -87,48 +79,78 @@ let
 
     lldb = callPackage ./lldb.nix {};
 
+    # Below, is the LLVM bootstrapping logic. It handles building a
```
Comment (TravisWhitaker) [inline pkgs/development/compilers/llvm/7/default.nix:82]: Hmm, why does this go in LLVM 7 and not the others?

Comment (Ericson2314) [reply pkgs/development/compilers/llvm/7/default.nix:82]: It was already in the others,

Comment (TravisWhitaker) [reply pkgs/development/compilers/llvm/7/default.nix:82]: ahhhh of course, sorry for the noise.

Comment (Ericson2314) [reply pkgs/development/compilers/llvm/7/default.nix:82]: no worries!

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -8181,14 +8180,6 @@ in
     stripped = false;
   }));
 
-  libstdcxxHook = makeSetupHook
```
Comment (matthewbauer) [inline pkgs/top-level/all-packages.nix:8184]: We may need a migration or warning for this.

It may not be actively used though

File: pkgs/build-support/cc-wrapper/default.nix [hunk 3]
```diff
@@ -267,6 +277,26 @@ stdenv.mkDerivation {
...
     '')
 
+    + ''
+      ##
+      ## General libc++ support
+      ##
+
+    ''
+    + optionalString isClang ''
+      echo "-isystem $out/include/c++/v1" >> $out/nix-support/libcxx-cxxflags
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix:287]: why is this needed with `isystem ${libcxx}/include/c++/v1` below?

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix:287]: also it appears that the /include/c++/v1 thing is only in libc++, not libstdc++

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix:287]: (although using libstdc++ with clang is probably unsupported)

File: pkgs/build-support/cc-wrapper/default.nix [hunk 4]
```diff
@@ -239,6 +231,24 @@ stdenv.mkDerivation {
...
+
+      echo "-target ${targetPlatform.config}" >> $out/nix-support/cc-cflags
+    ''
+
+    + optionalString (cc ? gcc && isClang && libcxx == null) ''
+      ##
+      ## GCC libs for non-GCC support
+      ##
+
+      echo "-B${cc.gcc.cc}/lib/gcc/${targetPlatform.config}/${cc.gcc.version}" >> $out/nix-support/cc-cflags
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix:247]: does `/lib/gcc/${targetPlatform.config}` exists on native builds?

File: pkgs/build-support/cc-wrapper/default.nix [hunk 5]
```diff
@@ -115,7 +109,7 @@ stdenv.mkDerivation {
     # Binutils, and Apple's "cctools"; "bintools" as an attempt to find an
     # unused middle-ground name that evokes both.
     inherit bintools;
-    inherit libc nativeTools nativeLibc nativePrefix isGNU isClang default_cxx_stdlib_compile;
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix:118]: This may be used outside by wrappers outside of nixpkgs. might be worth keeping, but just as meta-information.

File: pkgs/applications/networking/browsers/firefox/common.nix [hunk 6]
```diff
@@ -178,9 +178,9 @@ stdenv.mkDerivation ({
 
     BINDGEN_CFLAGS="$(< ${stdenv.cc}/nix-support/libc-cflags) \
```
Comment (matthewbauer) [inline pkgs/applications/networking/browsers/firefox/common.nix:179]: Maybe we should just expose all of these flags in the cc-wrapper ? Something like `stdenv.cc.cflags_compile` and `stdenv.cc.ld_flags`

Comment (matthewbauer) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: that is the command to get it since we don't know the actual values at evaluation

File: pkgs/os-specific/linux/firmware/fwupdate/default.nix [hunk 7]
```diff
@@ -13,7 +20,11 @@ in stdenv.mkDerivation {
     ./do-not-create-sharedstatedir.patch
   ];
 
-  NIX_CFLAGS_COMPILE = "-I${gnu-efi}/include/efi -Wno-error=address-of-packed-member";
+  NIX_CFLAGS_COMPILE = builtins.toString [
```
Comment (matthewbauer) [inline pkgs/os-specific/linux/firmware/fwupdate/default.nix:23]: i don't think toString is needed - nix does it for you when you put it into a drv

File: pkgs/build-support/cc-wrapper/default.nix (hunk 5 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:118]: It might be, but I think that's bad :)

File: pkgs/os-specific/linux/firmware/fwupdate/default.nix (hunk 7 above)
Comment (Ericson2314) [reply pkgs/os-specific/linux/firmware/fwupdate/default.nix:23]: True, I did this for sake of the upcoming structured attrs.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 4 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:247]: I think it does as of recently.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 3 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:287]: Isn't the other `$out/include/c++/v1`? I don't know what this is about, to be honest.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 6 above)
Comment (Ericson2314) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: It would be nice to uses more Nix so we did know the values. (e.g. have a nix boolean and then assert at run-time it was correct.)

File: pkgs/build-support/cc-wrapper/default.nix (hunk 3 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix]: Oh right, that was me getting confused on the setup hooks. This is indeed wrong.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 5 above)
Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix:118]: There's no good way to do use libraries that want these c flags in without some meta-info like this. For instance bindgen in rust or anything that uses libclang. If we're going to be removing it, we should at least have a replacement.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 6 above)
Comment (matthewbauer) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: I think that requires Nix-level setup-hooks to work correctly. I think that might be too ambitious for now.

Comment (matthewbauer) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: I'm thinking something like `/nix-support/get-cflags.sh` which does all of this for you.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 5 above)
Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix:118]: see https://github.com/NixOS/nixpkgs/pull/26474 for why this was exported

Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:118]: Right I would rather have a more general way to get flags and just this one in particular. The default part (i.e. miy get overriden) is no good too.

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 6 above)
Comment (Ericson2314) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: Well getting all the -I flags for other packaged requires that, but not just the std library stuff I think.

**Comment by matthewbauer** [comment]
I think we can also remove some of the old hacks like in https://github.com/NixOS/nixpkgs/pull/26709/files too

File: pkgs/applications/networking/browsers/firefox/common.nix (hunk 6 above)
Comment (matthewbauer) [reply pkgs/applications/networking/browsers/firefox/common.nix:179]: Right these should just need the basic c/c++ headers and library paths.

**Comment by LnL7** [comment]
Does that change here? I would expect this to still only include it for clang++. Come to think of it, wouldn't `CC=$CXX` be a nicer fix for those cases or does that not work?

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh [hunk 8]
```diff
@@ -127,11 +125,16 @@ if [ "$NIX_@infixSalt@_ENFORCE_NO_NATIVE" = 1 ]; then
...
+if [[ "$stdInclude" = 1 ]]; then
+    NIX_@infixSalt@_CFLAGS_COMPILE="$NIX_@infixSalt@_CSTDLIB_COMPILE $NIX_@infixSalt@_CFLAGS_COMPILE"
+fi
+NIX_@infixSalt@_CFLAGS_LINK="$NIX_@infixSalt@_CSTDLIB_LINK  $NIX_@infixSalt@_CFLAGS_LINK"
+
 if [[ "$isCpp" = 1 ]]; then
-    if [[ "$cppInclude" = 1 ]]; then
-        NIX_@infixSalt@_CFLAGS_COMPILE+=" ${NIX_@infixSalt@_CXXSTDLIB_COMPILE:-@default_cxx_stdlib_compile@}"
+    if [[ "$stdInclude" = 1 ]]; then
+        NIX_@infixSalt@_CFLAGS_COMPILE="$NIX_@infixSalt@_CXXSTDLIB_COMPILE $NIX_@infixSalt@_CFLAGS_COMPILE"
```
Comment (LnL7) [inline pkgs/build-support/cc-wrapper/cc-wrapper.sh:135]: Does the order matter here? I recall something about `+=` being much faster.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 9]
```diff
@@ -115,7 +109,7 @@ stdenv.mkDerivation {
...
+    inherit libc nativeTools nativeLibc nativePrefix isGNU isClang;
```
Comment (LnL7) [inline pkgs/build-support/cc-wrapper/default.nix:112]: Would it make sense to expose the libc++ of the compiler, similar to libc?

File: pkgs/build-support/cc-wrapper/add-flags.sh [hunk 10]
```diff
@@ -33,15 +41,27 @@ NIX_@infixSalt@_CFLAGS_COMPILE="-B@out@/bin/ $NIX_@infixSalt@_CFLAGS_COMPILE"
...
+if [ -e @out@/nix-support/libcxx-ldflags ]; then
+    NIX_@infixSalt@_CXXSTDLIB_LINK+=" $(< @out@/nix-support/libcxx-ldflags)"
 fi
 
 if [ -e @out@/nix-support/cc-cflags ]; then
     NIX_@infixSalt@_CFLAGS_COMPILE="$(< @out@/nix-support/cc-cflags) $NIX_@infixSalt@_CFLAGS_COMPILE"
 fi
 
+if [ -e @out@/nix-support/cc-cxxflags ]; then
+    NIX_@infixSalt@_CFLAGS_COMPILE="$(< @out@/nix-support/cc-cxxflags) $NIX_@infixSalt@_CFLAGS_COMPILE"
```
Comment (LnL7) [inline pkgs/build-support/cc-wrapper/add-flags.sh:60]: Why does this end up in CFLAGS instead of CXXFLAGS?

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh (hunk 8 above)
Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/cc-wrapper.sh:135]: Is `+=` much faster? If so, we probably should rewrite everything to use it.

**Comment by matthewbauer** [comment]
> Does that change here? I would expect this to still only include it for clang++. Come to think of it, wouldn't `CC=$CXX` be a nicer fix for those cases or does that not work?

Yeah, you're right - I was forgetting what the issue was there. We still need CXX to be used.

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh (hunk 8 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/cc-wrapper.sh:135]: The order does sometimes matter: wrapper-provided things should generally be overrideable by explicit arguments.

What I currently have causes a *segfault* during bootstrapping, so we are far from being able to see what `+=` we can get away with, sadly.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 9 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:112]: Yeah, though `null` for `libstdc++` is rather unfortunate. Oh well, all the more reason to get on splitting up the GCC derivation.

File: pkgs/build-support/cc-wrapper/add-flags.sh (hunk 10 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/add-flags.sh:60]: Good point, though I am not sure that file is made by anything just yet. Historically we didn't have a `CXXFLAGS_COMPILE` and I am not sure whether I should be adding one in this, actually. Seems like the right thing to do but maybe later so we have 1 less moving part.

I originally thought we could do `CXXSTDLIB` -> `CXXFLAGS`, but that is not actually the case as we have that tricky logic around `-nostdinc`.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (Mic92) [reply pkgs/top-level/all-packages.nix:8184]: aliases.nix should be enough.

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh (hunk 8 above)
Comment (Mic92) [reply pkgs/build-support/cc-wrapper/cc-wrapper.sh:135]:
```console
$ hyperfine "bash -c 'a=0; for i in \$(seq 1 100000); do a+=0; done'" "bash -c 'a=0; for i in \$(seq 1 100000); do a=\"0$a\"; done'"
Benchmark #1: bash -c 'a=0; for i in $(seq 1 100000); do a+=0; done'
  Time (mean ± σ):     983.5 ms ±   8.9 ms    [User: 494.8 ms, System: 491.3 ms]
  Range (min … max):   973.2 ms … 999.2 ms    10 runs

Benchmark #2: bash -c 'a=0; for i in $(seq 1 100000); do a="0"; done'
  Time (mean ± σ):     130.6 ms ±   2.3 ms    [User: 123.1 ms, System: 10.3 ms]
  Range (min … max):   127.3 ms … 135.8 ms    22 runs

Summary
  'bash -c 'a=0; for i in $(seq 1 100000); do a="0"; done'' ran
    7.53 ± 0.15 times faster than 'bash -c 'a=0; for i in $(seq 1 100000); do a+=0; done''
hyperfine "bash -c 'a=0; for i in \$(seq 1 100000); do a+=0; done'"   7.96s user 5.32s system 100% cpu 13.208 total
```

File: pkgs/build-support/cc-wrapper/default.nix (hunk 9 above)
Comment (LnL7) [reply pkgs/build-support/cc-wrapper/default.nix:112]: Perhaps for now `gcc` would make sense? Maybe that could also solve `cc ? gcc`.

File: pkgs/build-support/cc-wrapper/cc-wrapper.sh (hunk 8 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/cc-wrapper.sh]: This is converted back to the way it was.

File: pkgs/build-support/cc-wrapper/add-flags.sh (hunk 10 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/add-flags.sh]: There no longer is a `NIX_CXXFLAGS`

**Comment by Ericson2314** [comment]
OK with the last push I have both `stdenv`s building, so now we can clean up in earnest.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 9 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:112]: I'm on the fence about this---I do want to get rid of `cc.gcc` but it's a bit disingenous unless we can somehow filter the other libraries out. Maybe we can leave that for a follow-up PR?

**Comment by Mic92** [comment]
We also have some tools that consume the c/c++ standard includes: cquery, clang-tools and ccls

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix:8313]: I am not sure what should be aliased?

Comment (Mic92) [reply pkgs/top-level/all-packages.nix:8315]: Not aliased but we need a `throw` statement in `aliases.nix` that generates a warning.

Comment (Ericson2314) [reply pkgs/top-level/all-packages.nix:8315]: Ah! sounds good.

**Comment by Mic92** [comment]
> We also have some tools that consume the c/c++ standard includes: cquery, clang-tools and ccls

As it turns out those are not affected by this change. I removed cquery last week. Clangd and CCLS only take the libcxx include path as input. I remember to use `default_cxx_stdlib_compile` for youcompletme at some point, but this was more of a hack and using something like `${llvmPackages.libcxx}/include/c++/v1` would be cleaner.

**Comment by Ericson2314** [comment]
OK, it evals! I don't think we need any more back-compat stuff ATM, but happy to add things in requested.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 9 above)
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix:113]: https://github.com/NixOS/nixpkgs/pull/91293 got rid of `cc.gcc` here.
